Anonymization in/out

AJun01 opened the pull request:
### Summary

This pull request replaces https://github.com/flexivian/flexbench/pull/74. The Compare branch has been renamed from anonymization/v1.0 to AJun01/fix-issue-71 and renamed PR into Anonymization in/out.

### Changes

Renamed the branch to AJun01/fix-issue-71
Retained all commits and changes from the old branch

### Context

# Anonymization Feature v1.0

## Introduction

Releasing of the **Anonymization Feature v1.0**. This feature is designed to enhance the privacy and security of our users by filtering out sensitive information in generated HTTP requests. 

## Key Features

- **JSON Anonymization**
- **Form Data Anonymization**

## How It Works

The Anonymization Feature v1.0 utilizes advanced regular expressions to detect sensitive information in HTTP requests. When enabled using setAnonymization(), it processes the request body and replaces the detected sensitive information with anonymized placeholders.

### Example

**Original Request Body**:
```
{
  "email": "YujunLiu@example.com",
  "name": "Yujun Liu",
  "address": "22 Bond str"
}
```

**Anonymized Request Body**:
```
{
  "email": "************@email.com",
  "name": "*Anonymized* Name",
  "address": "*Anonymized* Address"
}
```

## How to Enable

To enable the Anonymization Feature, simply call the `anonymizeMode(true)` function in your configuration. This will ensure that all HTTP requests are processed to filter out sensitive information before being sent.

### Code Example

```javascript
var trafficSimulator = require('flexbench');

trafficSimulator.anonymizeMode(true); 


```

## Test(Anonymizer.js)

- File involved: 
<img width="187" alt="image" src="https://github.com/flexivian/flexbench/assets/94469665/8b8b924d-7098-4102-a1d4-32f02890b624">

- Run test:
```
npm test
```

- Result: **JSON**

<img width="500" alt="image" src="https://github.com/flexivian/flexbench/assets/94469665/2c675417-2ad3-43cb-909c-e7f7032c49e8">

- Result: **Form Data**

<img width="651" alt="image" src="https://github.com/flexivian/flexbench/assets/94469665/b4c5c476-6e67-43bc-9c58-9640f93f4196">

## Test(simple-request(anonymized).js)

- File involved:
<img width="212" alt="image" src="https://github.com/flexivian/flexbench/assets/94469665/46594706-89a2-44d9-a6b6-8459ee723df8">

- Run test:
```
node simple-request\(anonymized\).js
```
- Result:
<img width="1415" alt="image" src="https://github.com/flexivian/flexbench/assets/94469665/1c283910-b4b2-4e2a-b40d-fead45c45371">


## Yujun 6/16/2024 Created
## Yujun 7/4/2024 Updated

mariosk reviewed (commented):
@AJun01 please write in the documentation of the PR how you execute the tests. I checked out your branch and could not run the tests, got this error:

`
mkara@DESKTOP-MTFMFNO MINGW64 /d/repos/develop-flexbench (AJun01/fix-issue-71)
$ npm test

> flexbench@0.1.0 test
> mocha -R spec


 Exception during run: Error: Cannot find module './lib/anonymizer.js'
Require stack:
- D:\repos\develop-flexbench\lib\main.js
- D:\repos\develop-flexbench\test\counted-requests.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1212:15)
    at Module._load (node:internal/modules/cjs/loader:1038:27)
    at wrapModuleLoad (node:internal/modules/cjs/loader:212:19)
    at Module.require (node:internal/modules/cjs/loader:1297:12)
    at require (node:internal/modules/helpers:123:16)
    at Object.<anonymous> (D:\repos\develop-flexbench\lib\main.js:12:18)
    at Module._compile (node:internal/modules/cjs/loader:1460:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1544:10)
    at Module.load (node:internal/modules/cjs/loader:1275:32)
    at Module._load (node:internal/modules/cjs/loader:1091:12)
    at wrapModuleLoad (node:internal/modules/cjs/loader:212:19)
    at Module.require (node:internal/modules/cjs/loader:1297:12)
    at require (node:internal/modules/helpers:123:16)
    at Object.<anonymous> (D:\repos\develop-flexbench\test\counted-requests.js:3:24)
    at Module._compile (node:internal/modules/cjs/loader:1460:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1544:10)
    at Module.load (node:internal/modules/cjs/loader:1275:32)
    at Module._load (node:internal/modules/cjs/loader:1091:12)
    at wrapModuleLoad (node:internal/modules/cjs/loader:212:19)
    at cjsLoader (node:internal/modules/esm/translators:318:5)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:258:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:262:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:474:24)
    at async formattedImport (D:\repos\develop-flexbench\node_modules\mocha\lib\nodejs\esm-utils.js:9:14)
    at async exports.requireOrImport (D:\repos\develop-flexbench\node_modules\mocha\lib\nodejs\esm-utils.js:42:28)
    at async exports.loadFilesAsync (D:\repos\develop-flexbench\node_modules\mocha\lib\nodejs\esm-utils.js:100:20)
    at async singleRun (D:\repos\develop-flexbench\node_modules\mocha\lib\cli\run-helpers.js:158:3)
    at async exports.handler (D:\repos\develop-flexbench\node_modules\mocha\lib\cli\run.js:370:5) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'D:\\repos\\develop-flexbench\\lib\\main.js',
    'D:\\repos\\develop-flexbench\\test\\counted-requests.js'
  ]
}
`

iskitsas commented:
It is not clear how to test the new feature.

For example there is a run function here:
<img width="866" alt="image" src="https://github.com/flexivian/flexbench/assets/1267871/e04a1c98-49a4-4776-b68e-c96bcf211155">

and then a sample js file:
<img width="925" alt="image" src="https://github.com/flexivian/flexbench/assets/1267871/17a32658-c76d-44a4-8beb-5bc07c1f3522">

@AJun01 , do I need to call somehow `setAnonymization(true)` and then use the `anonymizeHttpResponse(response, responseBody)` function?

AJun01 commented:
I need to update the instructions on how to utilize this feature after I figure out how to integrate this into the main.js, the explanation so far is misleading.

AJun01 commented:
> It is not clear how to test the new feature.
> 
> For example there is a run function here: <img alt="image" width="866" src="https://private-user-images.githubusercontent.com/1267871/345624691-e04a1c98-49a4-4776-b68e-c96bcf211155.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxOTIxNTgsIm5iZiI6MTcyMDE5MTg1OCwicGF0aCI6Ii8xMjY3ODcxLzM0NTYyNDY5MS1lMDRhMWM5OC00OWE0LTQ3NzYtYjY4ZS1jOTZiY2YyMTExNTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDVUMTUwNDE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2ZhYWI0YjhlZTU2ZWRjY2I5NjRiMGE3NTNlMjliM2QyNTBjYWFhOTcxMmY0Mjk5MDkwOGY3Y2MxMjZkOWJmZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.lB71EmXzF1ACm8yFpCfrby8jH2I9As307ZqlBASQbUA">
> 
> and then a sample js file: <img alt="image" width="925" src="https://private-user-images.githubusercontent.com/1267871/345624928-17a32658-c76d-44a4-8beb-5bc07c1f3522.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxOTIxNTgsIm5iZiI6MTcyMDE5MTg1OCwicGF0aCI6Ii8xMjY3ODcxLzM0NTYyNDkyOC0xN2EzMjY1OC1jNzZkLTQ0YTQtOGJlYi01YmMwN2MxZjM1MjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDVUMTUwNDE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njc2MTEzNTQ5ODUyMzlhNjQwMTdkZDMwOTYwYzU3OGNmZThlNGEyZmZhN2YyODc3M2YyM2I4MzUyYTU1MGJjNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.v6llLh91xEQwG_XfSCuHAjSK_n12e29W85uqvO1Baxs">
> 
> @AJun01 , do I need to call somehow `setAnonymization(true)` and then use the `anonymizeHttpResponse(response, responseBody)` function?

I have updated the new PR Summary, and removed misleading content, please let me know if there is any problem

AJun01 commented:
> It is not clear how to test the new feature.
> 
> For example there is a run function here: <img alt="image" width="866" src="https://private-user-images.githubusercontent.com/1267871/345624691-e04a1c98-49a4-4776-b68e-c96bcf211155.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwODAxNTcsIm5iZiI6MTcyMTA3OTg1NywicGF0aCI6Ii8xMjY3ODcxLzM0NTYyNDY5MS1lMDRhMWM5OC00OWE0LTQ3NzYtYjY4ZS1jOTZiY2YyMTExNTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTVUMjE0NDE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTFkNDNiMTFkNzc3NGUyNTU0M2IzZDNlYmUyMGYzNDEwNDQxNjI4NDU3YWE5NTc3ZWYyYmNhNjY0N2Q5NTU0MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.IMgLvuj2nqTgd9dXlIPDLEsFZuFukPahCqGreFW4Gks">
> 
> and then a sample js file: <img alt="image" width="925" src="https://private-user-images.githubusercontent.com/1267871/345624928-17a32658-c76d-44a4-8beb-5bc07c1f3522.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwODAxNTcsIm5iZiI6MTcyMTA3OTg1NywicGF0aCI6Ii8xMjY3ODcxLzM0NTYyNDkyOC0xN2EzMjY1OC1jNzZkLTQ0YTQtOGJlYi01YmMwN2MxZjM1MjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTVUMjE0NDE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTU1ZjNkMWEwYzU4NTJjOTI0NDc4OTIyNGY4MDc5NzZmNzcwMGQ0ZWNmYjNhYjIyZDUwNGRkM2Y0MzJkMzk5ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7Tf9YLJTa8BXhLdE12jTqWAsDJLlCsT9KENIXYEQfOI">
> 
> @AJun01 , do I need to call somehow `setAnonymization(true)` and then use the `anonymizeHttpResponse(response, responseBody)` function?

you can run ```npm test```to run tests, which will run these two files that included the testing for each anonymizer functions.
<img width="187" alt="image" src="https://github.com/user-attachments/assets/40fca45a-3aa9-45de-8fe9-74836717e295"> 

Also, if you want to customize your testing, create new request just like other example request files but enable the anonymized mode by set ```trafficSimulator.anonymizeMode(true); ``` the traffic created will be anonymized, as well as the responses.

AJun01 commented on package.json:
```diff
@@ -6,10 +6,14 @@
   "dependencies": {
     "agentkeepalive": "^4.2.1",
     "express": "^4.17.3",
+    "flexbench": "^1.0.0",
```
before I know how to use npm link, I installed it just in case any errors when require('flexbench') in the example files, I will remove it.

mariosk replied:
This is fixed now.

AJun01 commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,109 @@
+const phoneRegex = /(\+?\d{1,4}[\s-]?)?(\(?\d{3}\)?[\s-]?)?[\d\s-]{7,10}/g;
+const urlRegex = /https?:\/\/[^\s/$.?#].[^\s]*/g;
+const stateRegex = /\b[A-Z]{2}\b/g;
+const zipRegex = /\b\d{5}(?:-\d{4})?\b/g;
```
adding it to string type body request

AJun01 commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,109 @@
+        return obj;
+    }
+
+    for (let key in obj) {
```
added the recursive call to handle nested loops

AJun01 commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,109 @@
+            } else if (Array.isArray(obj[key])) {
+                obj[key] = obj[key].map(item => anonymizeText(item));
+            } else {
+                if (key === 'street' || key === 'city' || key === 'country') {
```
added an anonymizeField to handle addresses explicitly also included case check

mariosk replied:
Fixed.

AJun01 commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,109 @@
+const addressRegex = /\b(\d{1,5}\s\w+(?:\s\w+)*(?:\s(Street|St|Avenue|Ave|Boulevard|Blvd|Road|Rd|Lane|Ln|Drive|Dr|Court|Ct|Circle|Cir))?)\b/g;
+const phoneRegex = /(\+?\d{1,4}[\s-]?)?(\(?\d{3}\)?[\s-]?)?[\d\s-]{7,10}/g;
+const urlRegex = /https?:\/\/[^\s/$.?#].[^\s]*/g;
+const stateRegex = /\b[A-Z]{2}\b/g;
```
adding it to the string type filter

AJun01 commented on lib/main.js:
```diff
@@ -56,6 +57,9 @@ var timers=[];
 var maxDelay;//ms
 var isRandomDelay=false;
+
+var anonymizationEnabled = gh.args(true);
```
since the value set to be true as default, I am removing this

AJun01 commented on lib/main.js:
```diff
@@ -76,6 +80,11 @@ var EVENT={
     error:'error'
 }
 
+function anonymizeMode(flag) {
```
this is the setter for anonymizeMode where developers can turn this feature on or off

AJun01 commented on lib/main.js:
```diff
@@ -417,19 +431,30 @@ function request(options,cb){
+                responseBody = anonymizer.anonymizeText(responseBody);
+            }
+            if (cb) {
+                cb(response, responseBody);
```
fixing it

AJun01 replied:
added an if statement for remaining compatibility while anonymizing

AJun01 replied:
I will keep avoiding refactoring of code structures

iskitsas commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,276 @@
+    }
+
+    text = anonymizeEmail(text);
+    test = anonymizeAddress(test);
```
What is this `test` variable, typo?

AJun01 replied:
typo fixed

iskitsas commented on lib/main.js:
```diff
@@ -76,6 +78,11 @@ var EVENT={
 }
 
+function anonymizeMode(flag) {
+    anonymizationEnabled = flag;
```
The variable `anonymizationEnabled` should not be under the scope of main.js file. You can encapsulate the anonymization toggle(true|false) inside the anonymizer.js, and where needed, to call a function from anonymizer.js to get if the anonymization is on or off. eg. anonymizer.isOn could return true or false

AJun01 replied:
ok

iskitsas commented on lib/main.js:
```diff
@@ -385,54 +392,192 @@ function createPipelinedRequest(cb){
+    } else {
+        _options = options;
+    }
+//--------------------------Anonymizating Request----------------------------------------------------------------------------------
```
All this code of anonymizing request should not be added under main.js. You have to keep the changes in main.lib as minimum as possible and pass the responsibility to the anonymizer lib.

Check my previous comment on the anonymizationEnabled variable, and move all this new code under anonymizer.js with proper refactoring.

I would expect something like this:
```
if(anonymizer.isOn){
  anonymizer.startProcess(options, ..);
}
```

iskitsas replied:
Additionally, you are using excessively console.log. Are you able to add breakpoints and debug properly through your IDE? Which IDE are you using?

Remove all those console.log, and keep them only for informative purposes that give value to a developer

AJun01 replied:
For the IDE, I am using VS Code, I used debugging feature for other IDEs before but never tried to use it at VS Code with javascript, I will adapt to use it, thank you for point out.

iskitsas commented on lib/main.js:
```diff
@@ -385,54 +392,192 @@ function createPipelinedRequest(cb){
     }
+
+
+    //--------------------------Anonymizating Response----------------------------------------------------------------------------------
```
All the code added here must be part of anonymizer.js. You have to rework this part, and keep the initial implementation of the main.js intact with just the call to anonymizer function that would anonymize the response.

I would expect something like this:

Existing code with minor intervention:
```
var req = protocol.request(_options, function (response) {
        response.on('end', function () {
            incrCounter(response.statusCode);
        });

        if(cb){
            cb(!anonymizer.isOn ? response : anonymizer.manipulateResponse(response));
        }
        response.resume();
    });
```

You can modify(anonymize) the response data before sending it to the client, so you do not break any existing usages of the existing library.

AJun01 replied:
looks good! I will change it now

iskitsas commented on lib/main.js:
```diff
@@ -385,54 +392,192 @@ function createPipelinedRequest(cb){
+                if (cb) {
+                    console.log('Invoking callback with response and parsedResponse');
+                   
+                    if (cb.length === 2) {
```
This code would not be "backwards compatible" with our main lib. See my comment at L474

iskitsas commented on test/anonymized-Form-test.js:
As stated in previous comment, apart from kebab-case, rename the filename in lowercase

iskitsas commented on test/anonymized-JSON-test.js:
As stated in previous comment, apart from kebab-case, rename the filename in lowercase

iskitsas commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,571 @@
+    events = require('events'),
+    exec = require('child_process').exec,
+    gh = require('./lib-helper.js'),
+    proxy = require('./ThrottleProxy/throttleServer.js'),
```
Why do you need ThrottleProxy in anonymizer?

AJun01 replied:
I accidentally replaced the code of anonymizer with main.js, I am recovering it now

iskitsas commented on lib/main.js:
```diff
@@ -385,54 +387,61 @@ function createPipelinedRequest(cb){
-        }
+        response.on('parsedBody', function (parsedBody) { 
+            if (cb) {
+                cb(!anonymizer.isAnonymizationEnabled() ? response : parsedBody);
```
The response should always be returned in cb

iskitsas commented on test/anonymized-Form-test.js:
Filename convention is: lowercase and kebab-case, do the required changes

iskitsas commented on test/anonymized-JSON-test.js:
Filename convention is: lowercase and kebab-case, do the required changes

iskitsas commented on lib/anonymizer.js:
This file is messed, check your previous commits

iskitsas commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,338 @@
+                //urlencoded
+                case contentType.includes('application/x-www-form-urlencoded'):
+                    options.body = anonymizeUrlEncoded(options.body);
+                    //console.log('Anonymized URL-encoded(Request):', options.body);
```
Remove all commented out console logs

iskitsas commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,338 @@
+                //json
+                case contentType.includes('application/json'):
+                    options.body = JSON.stringify(anonymizeObject(JSON.parse(options.body)));
+                    //console.log('Anonymized JSON(Request):', options.body);
```
Remove all commented out console logs

iskitsas commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,338 @@
+                    anonymizeFormData(options.body, (err, anonymizedData) => {
+                        if (!err) {
+                            options.body = anonymizedData;
+                            //console.log('Anonymized Multipart Form-data(Request):', options.body);
```
Remove all commented out console logs

iskitsas commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,338 @@
+                //default set as string
+                default:
+                    options.body = anonymizeText(options.body);
+                   // console.log('Anonymized Other Type(Request):', options.body);
```
Remove all commented out console logs

AJun01 commented:
console logs removed

iskitsas commented on doc/flexbench/docs/GSoC/OpenAPI-Integration-Report.md:
This file is irrelevant with this PR

iskitsas commented on lib/anonymizer.js:
```diff
@@ -263,55 +256,44 @@ function anonymizeResponse(response, cb) {
     let responseBody = '';
     response.on('data', function (chunk) {
-        console.log('Received chunk:', chunk.toString()); // Added logging here
+        console.log('Received chunk:', chunk.toString()); 
```
There is no need to log body chunks

AJun01 commented:
> Trying to run tests locally, after following the instructions of this PR I get this error:
> 
> ```
> iskitsas001@GR_D36WYRX1MV flexbench % npm test   
> 
> > flexbench@1.0.0 test
> > mocha -R spec
> 
> 
>  Exception during run: Error: Cannot find module 'cheerio'
> Require stack:
> - /Users/iskitsas001/gitRepos/flexivian/flexbench/lib/anonymizer.js
> - /Users/iskitsas001/gitRepos/flexivian/flexbench/test/anonymized-form-test.js
>     at Module._resolveFilename (node:internal/modules/cjs/loader:1070:15)
>     at Module._load (node:internal/modules/cjs/loader:923:27)
>     at Module.require (node:internal/modules/cjs/loader:1137:19)
> ```
> 
> How to reproduce:
> 
> 1. Checkout this PR
> 2. Navigate to root folder
> 3. `npm install`
> 4. `npm test`
> 
> After installing the missing dependency manually, and re-running `npm test` I get this error:
> 
> ```
> iskitsas001@GR_D36WYRX1MV flexbench % npm test
> 
> > flexbench@1.0.0 test
> > mocha -R spec
> 
> 
>  Exception during run: Error: ENOENT: no such file or directory, open '/Users/iskitsas001/gitRepos/flexivian/flexbench/lib/anonymizerConfig.json'
>     at Object.openSync (node:fs:592:3)
>     at Object.readFileSync (node:fs:460:35)
>     at Object.<anonymous> (/Users/iskitsas001/gitRepos/flexivian/flexbench/lib/anonymizer.js:8:43)
>     at Module._compile (node:internal/modules/cjs/loader:1255:14)
>     at Module._extensions..js (node:internal/modules/cjs/loader:1309:10)
>     at Module.load (node:internal/modules/cjs/loader:1113:32)
>     at Module._load (node:internal/modules/cjs/loader:960:12)
>     at Module.require (node:internal/modules/cjs/loader:1137:19)
>     at require (node:internal/modules/helpers:121:18)
>     at Object.<anonymous> (/Users/iskitsas001/gitRepos/flexivian/flexbench/test/anonymized-form-test.js:3:47)
>     at Module._compile (node:internal/modules/cjs/loader:1255:14)
>     at Module._extensions..js (node:internal/modules/cjs/loader:1309:10)
>     at Module.load (node:internal/modules/cjs/loader:1113:32)
>     at Module._load (node:internal/modules/cjs/loader:960:12)
>     at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:165:29)
>     at ModuleJob.run (node:internal/modules/esm/module_job:192:25) {
>   errno: -2,
>   syscall: 'open',
>   code: 'ENOENT',
>   path: '/Users/iskitsas001/gitRepos/flexivian/flexbench/lib/anonymizerConfig.json'
> }
> ```

this error is caused by the fact that missing anonymizerConfig file is missing from this branch, which is included in PR #92. I have blocked the part where calling this file, now should be working.

iskitsas commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,320 @@
+const fs = require('fs');
+const path = require('path');
+
+// load anonymizerConfig
```
Remove commented out code if not used

mariosk replied:
This code is still there @AJun01, please remove it.

iskitsas commented on doc/flexbench/docs/GSoC/OpenAPI-Integration-Report.md:
This file is irrelevant with this PR, please remove

iskitsas commented:
When trying to run the test `simple-request-anonymized.js` I get this error:

```
iskitsas001@GR_D36WYRX1MV example % node simple-request-anonymized.js
/Users/iskitsas001/gitRepos/flexivian/flexbench/example/simple-request-anonymized.js:17
    trafficSimulator.anonymizeMode(true);
                     ^

TypeError: trafficSimulator.anonymizeMode is not a function
    at runTest (/Users/iskitsas001/gitRepos/flexivian/flexbench/example/simple-request-anonymized.js:17:22)
    at Object.<anonymous> (/Users/iskitsas001/gitRepos/flexivian/flexbench/example/simple-request-anonymized.js:79:1)
    at Module._compile (node:internal/modules/cjs/loader:1255:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1309:10)
    at Module.load (node:internal/modules/cjs/loader:1113:32)
    at Module._load (node:internal/modules/cjs/loader:960:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12)
    at node:internal/main/run_main_module:23:47

Node.js v20.2.0

```

iskitsas commented:
@AJun01 update the main README of the project to include how to run/use the simple-request-anonymized.js test

AJun01 commented:
> node simple-request-anonymized.js

since I have removed lots of code in main.js, I need add code to require('anonymizer') to this the simple-request-anonymized.js: anonymization.setAnonymization(true); changing it now

iskitsas commented on example/simple-request-anonymized.js:
```diff
@@ -0,0 +1,80 @@
+ * Scenario:
+ * Generate requests towards specific domain
+ **/
+var trafficSimulator = require('../lib/main');
```
This should not be changed

iskitsas commented on example/simple-request-anonymized.js:
```diff
@@ -0,0 +1,80 @@
+ * Generate requests towards specific domain
+ **/
+var trafficSimulator = require('../lib/main');
+var anonymizer = require('../lib/anonymizer');
```
The anonymizer dependency should not be imported here. You anonymizer feature should be exposed through the main library

iskitsas commented on README.md:
````diff
@@ -88,10 +88,33 @@ If you want to uninstall just unlink it like this:
+After you cloned your exmaple, at the top of the file, please import anonymizer.js from the repository:
+
+```
+var anonymizer = require('../lib/anonymizer');
````
This is not correct

iskitsas commented on example/simple-request-anonymized.js:
@AJun01 what are the exact steps for someone who is cloning the project for first time to run this test in his pc?

mariosk replied:
I suspect `PS D:\repos\develop-flexbench\example> node simple-request-anonymized.js`, right @AJun01?

iskitsas commented on lib/anonymizer.js:
```diff
@@ -0,0 +1,324 @@
+const passwordRegex = /(?<=name="password"[\s\S]*?\r\n\r\n)([A-Za-z0-9]+[A-Za-z0-9]*|[A-Za-z0-9]*[A-Za-z0-9]+)/g;
+
+//default
+let anonymizationEnabled = true;
```
@AJun01 anonymization should be not enabled by default. For example if I run the simple-request.js test from examples, despite having NOT set `trafficSimulator.setAnonymization(true);` , the feature is on.

iskitsas commented on lib/main.js:
```diff
@@ -385,54 +387,59 @@ function createPipelinedRequest(cb){
+            if(anonymizer.isAnonymizationEnabled)
+            anonymizer.anonymizeResponse(response, cb);
+        } else {
+            response.resume();
```
@AJun01 You have missed the cb(response) on the flow where anonymization is not enabled. Additionally, always use { } in if statements, even if it is one liner

iskitsas replied:
I would expect to see something like this, based on your code:

```
if (cb) {
  if(anonymizer.isAnonymizationEnabled) {
    anonymizer.anonymizeResponse(response, cb);
   }else {
     cb(response)
   }
}
response.resume();
```

iskitsas commented:
@AJun01 I am proposing some changes, see PR #94 
You can cherry pick the commit 4fa42b457e82022251602fb5051be6d3058f5521 if you want and continue from there (or just adopt the changes through a normal commit)

Let me know if you have any comments

mariosk commented:
> node simple-request-anonymized.js

@iskitsas since official flexbench package does not support this functionality yet, the error is normal. If you want to make it work locally you may run `npm link flexbench` under your root project folder and retry.

mariosk commented:
@AJun01 please fix the `npm test` failures:

```shell
PS D:\repos\develop-flexbench> npm test

> flexbench@1.0.0 test
> mocha -R spec       



  Anonymization Tests for Form Data  
PassThrough {
  _events: {
    close: undefined,
    error: undefined,
    prefinish: [Function: prefinish],
    finish: undefined,
    drain: undefined,
    data: undefined,
    end: undefined,
    readable: undefined
  },
  _readableState: ReadableState {
    highWaterMark: 16384,
    buffer: [
      <Buffer 2d 2d 62 6f 75 6e 64 61 72 79 31 32 33 0d 0a 43 6f 6e 74 65 6e 74 2d 44 69 73 70 6f 73 69 74 69 6f 6e 3a 20 66 6f 72 6d 2d 64 61 74 61 3b 20 6e 61 6d ... 436 more bytes>
    ],
    bufferIndex: 0,
    length: 486,
    pipes: [],
    awaitDrainWriters: null,
    [Symbol(kState)]: 1311500
  },
  _writableState: WritableState {
    highWaterMark: 16384,
    length: 0,
    corked: 0,
    onwrite: [Function: bound onwrite],
    writelen: 0,
    bufferedIndex: 0,
    pendingcb: 1,
    [Symbol(kState)]: 1091458316,
    [Symbol(kBufferedValue)]: null
  },
  allowHalfOpen: true,
  _maxListeners: undefined,
  _eventsCount: 1,
  headers: { 'content-type': 'multipart/form-data; boundary=boundary123' },
  [Symbol(shapeMode)]: true,
  [Symbol(kCapture)]: false,
  [Symbol(kCallback)]: null
}
    1) should anonymize form data

  Anonymization Tests for JSON Data
    2) should anonymize the email
    3) should anonymize the name
    4) should anonymize the street address
    5) should anonymize the city
    6) should anonymize the state
    7) should anonymize the zip code
    8) should anonymize the phone number
    9) should anonymize URLs
    ✔ should respect the feature flag

  TrafficSimulator - counted Requests
    #start()


  Anonymization Tests for Form Data
PassThrough {
  _events: {
    close: undefined,
    error: undefined,
    prefinish: [Function: prefinish],
    finish: undefined,
    drain: undefined,
    data: undefined,
    end: undefined,
    readable: undefined
  },
  _readableState: ReadableState {
    highWaterMark: 16384,
    buffer: [
      <Buffer 2d 2d 62 6f 75 6e 64 61 72 79 31 32 33 0d 0a 43 6f 6e 74 65 6e 74 2d 44 69 73 70 6f 73 69 74 69 6f 6e 3a 20 66 6f 72 6d 2d 64 61 74 61 3b 20 6e 61 6d ... 436 more bytes>
    ],
    bufferIndex: 0,
    length: 486,
    pipes: [],
    awaitDrainWriters: null,
    [Symbol(kState)]: 1311500
  },
  _writableState: WritableState {
    highWaterMark: 16384,
    length: 0,
    corked: 0,
    onwrite: [Function: bound onwrite],
    writelen: 0,
    bufferedIndex: 0,
    pendingcb: 1,
    [Symbol(kState)]: 1091458316,
    [Symbol(kBufferedValue)]: null
  },
  allowHalfOpen: true,
  _maxListeners: undefined,
  _eventsCount: 1,
  headers: { 'content-type': 'multipart/form-data; boundary=boundary123' },
  [Symbol(shapeMode)]: true,
  [Symbol(kCapture)]: false,
  [Symbol(kCallback)]: null
}
    1) should anonymize form data

  Anonymization Tests for JSON Data
    2) should anonymize the email
    3) should anonymize the name
    4) should anonymize the street address
    5) should anonymize the city
    6) should anonymize the state
    7) should anonymize the zip code
    8) should anonymize the phone number
    9) should anonymize URLs
    ✔ should respect the feature flag

  TrafficSimulator - counted Requests
    #start()
Response stream closed
Response stream closed
Response stream closed
Response stream closed
Response stream closed
      ✔ should create some requests (3081ms)

  Array
    #indexOf()
      ✔ should return -1 when the value is not present

  Dummy Tester
    #increaseCounterBy1()
      ✔ should return 26 when the 25 value is increased by 1


  4 passing (3s)
  9 failing

  1) Anonymization Tests for Form Data
       should anonymize form data:
     TypeError: Cannot read properties of undefined (reading 'first_name')
      at D:\repos\develop-flexbench\test\anonymized-form-test.js:22:35
      at anonymizeFormData (lib\anonymizer.js:187:16)
      at Context.<anonymous> (test\anonymized-form-test.js:17:9)
      at process.processImmediate (node:internal/timers:478:21)

  2) Anonymization Tests for JSON Data
       should anonymize the email:

      AssertionError [ERR_ASSERTION]: expected 'YujunLiu@example.com' to be 'anonymized_email@example.com'
      + expected - actual

      -YujunLiu@example.com
      +anonymized_email@example.com

      at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
      at Context.<anonymous> (test\anonymized-json-test.js:65:37)
      at process.processImmediate (node:internal/timers:478:21)

  3) Anonymization Tests for JSON Data
       should anonymize the name:

      AssertionError [ERR_ASSERTION]: expected 'Yujun Liu' to be 'Anonymized Name'
      + expected - actual

      -Yujun Liu
      +Anonymized Name

      at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
      at Context.<anonymous> (test\anonymized-json-test.js:70:36)
      at process.processImmediate (node:internal/timers:478:21)

  4) Anonymization Tests for JSON Data
       should anonymize the street address:

      AssertionError [ERR_ASSERTION]: expected '22 Bond St' to be 'Anonymized Address'
      + expected - actual

      -22 Bond St
      +Anonymized Address

      at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
      at Context.<anonymous> (test\anonymized-json-test.js:75:46)
      at process.processImmediate (node:internal/timers:478:21)

  5) Anonymization Tests for JSON Data
       should anonymize the city:

      AssertionError [ERR_ASSERTION]: expected 'New York' to be 'Anonymized Address'
      + expected - actual

      -New York
      +Anonymized Address

      at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
      at Context.<anonymous> (test\anonymized-json-test.js:80:44)
      at process.processImmediate (node:internal/timers:478:21)

  6) Anonymization Tests for JSON Data
       should anonymize the state:

      AssertionError [ERR_ASSERTION]: expected 'NY' to be 'XX'
      + expected - actual

      -NY
      +XX

      at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
      at Context.<anonymous> (test\anonymized-json-test.js:85:45)
      at process.processImmediate (node:internal/timers:478:21)

  7) Anonymization Tests for JSON Data
       should anonymize the zip code:

      AssertionError [ERR_ASSERTION]: expected '10012' to be '00000'
      + expected - actual

      -10012
      +00000

      at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
      at Context.<anonymous> (test\anonymized-json-test.js:90:43)
      at process.processImmediate (node:internal/timers:478:21)

  8) Anonymization Tests for JSON Data
       should anonymize the phone number:

      AssertionError [ERR_ASSERTION]: expected '+1-555-123-4567' to be '000-000-0000'
      + expected - actual

      -+1-555-123-4567
      +000-000-0000

      at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
      at Context.<anonymous> (test\anonymized-json-test.js:95:37)
      at process.processImmediate (node:internal/timers:478:21)

  9) Anonymization Tests for JSON Data
       should anonymize URLs:

      AssertionError [ERR_ASSERTION]: expected 'https://linkedin.com/in/yujunliu' to be 'https://anonymized-url.com'
      + expected - actual

      -https://linkedin.com/in/yujunliu
      +https://anonymized-url.com

      at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
      at Context.<anonymous> (test\anonymized-json-test.js:100:53)
      at process.processImmediate (node:internal/timers:478:21)

PS D:\repos\develop-flexbench>
```

AJun01 commented:
> @AJun01 please fix the `npm test` failures:
> 
> ```shell
> PS D:\repos\develop-flexbench> npm test
> 
> > flexbench@1.0.0 test
> > mocha -R spec       
> 
> 
> 
>   Anonymization Tests for Form Data  
> PassThrough {
>   _events: {
>     close: undefined,
>     error: undefined,
>     prefinish: [Function: prefinish],
>     finish: undefined,
>     drain: undefined,
>     data: undefined,
>     end: undefined,
>     readable: undefined
>   },
>   _readableState: ReadableState {
>     highWaterMark: 16384,
>     buffer: [
>       <Buffer 2d 2d 62 6f 75 6e 64 61 72 79 31 32 33 0d 0a 43 6f 6e 74 65 6e 74 2d 44 69 73 70 6f 73 69 74 69 6f 6e 3a 20 66 6f 72 6d 2d 64 61 74 61 3b 20 6e 61 6d ... 436 more bytes>
>     ],
>     bufferIndex: 0,
>     length: 486,
>     pipes: [],
>     awaitDrainWriters: null,
>     [Symbol(kState)]: 1311500
>   },
>   _writableState: WritableState {
>     highWaterMark: 16384,
>     length: 0,
>     corked: 0,
>     onwrite: [Function: bound onwrite],
>     writelen: 0,
>     bufferedIndex: 0,
>     pendingcb: 1,
>     [Symbol(kState)]: 1091458316,
>     [Symbol(kBufferedValue)]: null
>   },
>   allowHalfOpen: true,
>   _maxListeners: undefined,
>   _eventsCount: 1,
>   headers: { 'content-type': 'multipart/form-data; boundary=boundary123' },
>   [Symbol(shapeMode)]: true,
>   [Symbol(kCapture)]: false,
>   [Symbol(kCallback)]: null
> }
>     1) should anonymize form data
> 
>   Anonymization Tests for JSON Data
>     2) should anonymize the email
>     3) should anonymize the name
>     4) should anonymize the street address
>     5) should anonymize the city
>     6) should anonymize the state
>     7) should anonymize the zip code
>     8) should anonymize the phone number
>     9) should anonymize URLs
>     ✔ should respect the feature flag
> 
>   TrafficSimulator - counted Requests
>     #start()
> 
> 
>   Anonymization Tests for Form Data
> PassThrough {
>   _events: {
>     close: undefined,
>     error: undefined,
>     prefinish: [Function: prefinish],
>     finish: undefined,
>     drain: undefined,
>     data: undefined,
>     end: undefined,
>     readable: undefined
>   },
>   _readableState: ReadableState {
>     highWaterMark: 16384,
>     buffer: [
>       <Buffer 2d 2d 62 6f 75 6e 64 61 72 79 31 32 33 0d 0a 43 6f 6e 74 65 6e 74 2d 44 69 73 70 6f 73 69 74 69 6f 6e 3a 20 66 6f 72 6d 2d 64 61 74 61 3b 20 6e 61 6d ... 436 more bytes>
>     ],
>     bufferIndex: 0,
>     length: 486,
>     pipes: [],
>     awaitDrainWriters: null,
>     [Symbol(kState)]: 1311500
>   },
>   _writableState: WritableState {
>     highWaterMark: 16384,
>     length: 0,
>     corked: 0,
>     onwrite: [Function: bound onwrite],
>     writelen: 0,
>     bufferedIndex: 0,
>     pendingcb: 1,
>     [Symbol(kState)]: 1091458316,
>     [Symbol(kBufferedValue)]: null
>   },
>   allowHalfOpen: true,
>   _maxListeners: undefined,
>   _eventsCount: 1,
>   headers: { 'content-type': 'multipart/form-data; boundary=boundary123' },
>   [Symbol(shapeMode)]: true,
>   [Symbol(kCapture)]: false,
>   [Symbol(kCallback)]: null
> }
>     1) should anonymize form data
> 
>   Anonymization Tests for JSON Data
>     2) should anonymize the email
>     3) should anonymize the name
>     4) should anonymize the street address
>     5) should anonymize the city
>     6) should anonymize the state
>     7) should anonymize the zip code
>     8) should anonymize the phone number
>     9) should anonymize URLs
>     ✔ should respect the feature flag
> 
>   TrafficSimulator - counted Requests
>     #start()
> Response stream closed
> Response stream closed
> Response stream closed
> Response stream closed
> Response stream closed
>       ✔ should create some requests (3081ms)
> 
>   Array
>     #indexOf()
>       ✔ should return -1 when the value is not present
> 
>   Dummy Tester
>     #increaseCounterBy1()
>       ✔ should return 26 when the 25 value is increased by 1
> 
> 
>   4 passing (3s)
>   9 failing
> 
>   1) Anonymization Tests for Form Data
>        should anonymize form data:
>      TypeError: Cannot read properties of undefined (reading 'first_name')
>       at D:\repos\develop-flexbench\test\anonymized-form-test.js:22:35
>       at anonymizeFormData (lib\anonymizer.js:187:16)
>       at Context.<anonymous> (test\anonymized-form-test.js:17:9)
>       at process.processImmediate (node:internal/timers:478:21)
> 
>   2) Anonymization Tests for JSON Data
>        should anonymize the email:
> 
>       AssertionError [ERR_ASSERTION]: expected 'YujunLiu@example.com' to be 'anonymized_email@example.com'
>       + expected - actual
> 
>       -YujunLiu@example.com
>       +anonymized_email@example.com
> 
>       at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
>       at Context.<anonymous> (test\anonymized-json-test.js:65:37)
>       at process.processImmediate (node:internal/timers:478:21)
> 
>   3) Anonymization Tests for JSON Data
>        should anonymize the name:
> 
>       AssertionError [ERR_ASSERTION]: expected 'Yujun Liu' to be 'Anonymized Name'
>       + expected - actual
> 
>       -Yujun Liu
>       +Anonymized Name
> 
>       at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
>       at Context.<anonymous> (test\anonymized-json-test.js:70:36)
>       at process.processImmediate (node:internal/timers:478:21)
> 
>   4) Anonymization Tests for JSON Data
>        should anonymize the street address:
> 
>       AssertionError [ERR_ASSERTION]: expected '22 Bond St' to be 'Anonymized Address'
>       + expected - actual
> 
>       -22 Bond St
>       +Anonymized Address
> 
>       at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
>       at Context.<anonymous> (test\anonymized-json-test.js:75:46)
>       at process.processImmediate (node:internal/timers:478:21)
> 
>   5) Anonymization Tests for JSON Data
>        should anonymize the city:
> 
>       AssertionError [ERR_ASSERTION]: expected 'New York' to be 'Anonymized Address'
>       + expected - actual
> 
>       -New York
>       +Anonymized Address
> 
>       at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
>       at Context.<anonymous> (test\anonymized-json-test.js:80:44)
>       at process.processImmediate (node:internal/timers:478:21)
> 
>   6) Anonymization Tests for JSON Data
>        should anonymize the state:
> 
>       AssertionError [ERR_ASSERTION]: expected 'NY' to be 'XX'
>       + expected - actual
> 
>       -NY
>       +XX
> 
>       at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
>       at Context.<anonymous> (test\anonymized-json-test.js:85:45)
>       at process.processImmediate (node:internal/timers:478:21)
> 
>   7) Anonymization Tests for JSON Data
>        should anonymize the zip code:
> 
>       AssertionError [ERR_ASSERTION]: expected '10012' to be '00000'
>       + expected - actual
> 
>       -10012
>       +00000
> 
>       at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
>       at Context.<anonymous> (test\anonymized-json-test.js:90:43)
>       at process.processImmediate (node:internal/timers:478:21)
> 
>   8) Anonymization Tests for JSON Data
>        should anonymize the phone number:
> 
>       AssertionError [ERR_ASSERTION]: expected '+1-555-123-4567' to be '000-000-0000'
>       + expected - actual
> 
>       -+1-555-123-4567
>       +000-000-0000
> 
>       at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
>       at Context.<anonymous> (test\anonymized-json-test.js:95:37)
>       at process.processImmediate (node:internal/timers:478:21)
> 
>   9) Anonymization Tests for JSON Data
>        should anonymize URLs:
> 
>       AssertionError [ERR_ASSERTION]: expected 'https://linkedin.com/in/yujunliu' to be 'https://anonymized-url.com'
>       + expected - actual
> 
>       -https://linkedin.com/in/yujunliu
>       +https://anonymized-url.com
> 
>       at prop.<computed> [as exactly] (node_modules\should\lib\should.js:60:14)
>       at Context.<anonymous> (test\anonymized-json-test.js:100:53)
>       at process.processImmediate (node:internal/timers:478:21)
> 
> PS D:\repos\develop-flexbench>
> ```

did you set 
```
let anonymizationEnabled = false;
``` 
into true?

mariosk reviewed (approved):
Approved
